diff -Naur gearvr_kernel_removed/drivers/hid/hid-core.c gearvr_kernel/drivers/hid/hid-core.c
--- gearvr_kernel_removed/drivers/hid/hid-core.c	2016-04-10 00:16:40.362692000 -0500
+++ gearvr_kernel/drivers/hid/hid-core.c	2016-04-10 00:10:12.602421100 -0500
@@ -1834,6 +1834,17 @@
 	{ HID_BLUETOOTH_DEVICE(USB_VENDOR_ID_MICROSOFT, USB_DEVICE_ID_MS_PRESENTER_8K_BT) },
 	{ HID_BLUETOOTH_DEVICE(USB_VENDOR_ID_NINTENDO, USB_DEVICE_ID_NINTENDO_WIIMOTE) },
 	{ HID_BLUETOOTH_DEVICE(USB_VENDOR_ID_NINTENDO, USB_DEVICE_ID_NINTENDO_WIIMOTE2) },
+#if IS_ENABLED(CONFIG_HID_OVR)
+	{ HID_USB_DEVICE(USB_VENDOR_ID_OVR, USB_DEVICE_ID_OVR_TRACKER) },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_OVR, USB_DEVICE_ID_OVR_KTRACKER) },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_OVR, USB_DEVICE_ID_OVR_LATENCY_TESTER) },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_SAMSUNG_ELECTRONICS, USB_DEVICE_ID_SAMSUNG_GEARVR_1) },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_SAMSUNG_ELECTRONICS, USB_DEVICE_ID_SAMSUNG_GEARVR_2) },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_SAMSUNG_ELECTRONICS, USB_DEVICE_ID_SAMSUNG_GEARVR_3) },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_SAMSUNG_ELECTRONICS, USB_DEVICE_ID_SAMSUNG_GEARVR_4) },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_SAMSUNG_ELECTRONICS, USB_DEVICE_ID_SAMSUNG_GEARVR_5) },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_SAMSUNG_ELECTRONICS, USB_DEVICE_ID_SAMSUNG_GEARVR_6) },
+#endif
 	{ }
 };
 
diff -Naur gearvr_kernel_removed/drivers/hid/hid-ids.h gearvr_kernel/drivers/hid/hid-ids.h
--- gearvr_kernel_removed/drivers/hid/hid-ids.h	2016-04-10 00:15:41.960579300 -0500
+++ gearvr_kernel/drivers/hid/hid-ids.h	2016-04-10 00:10:12.617899400 -0500
@@ -929,4 +929,18 @@
 #define USB_VENDOR_ID_SIS	0x0457
 #define USB_DEVICE_ID_SIS_TS	0x1013
 
+#ifdef CONFIG_HID_OVR
+#define USB_VENDOR_ID_OVR		0x2833
+#define USB_DEVICE_ID_OVR_TRACKER	0x0001
+#define USB_DEVICE_ID_OVR_KTRACKER		0x0010
+#define USB_DEVICE_ID_OVR_LATENCY_TESTER	0x0101
+#endif
+
+#define USB_DEVICE_ID_SAMSUNG_GEARVR_1	0xa500
+#define USB_DEVICE_ID_SAMSUNG_GEARVR_2	0xa501
+#define USB_DEVICE_ID_SAMSUNG_GEARVR_3	0xa502
+#define USB_DEVICE_ID_SAMSUNG_GEARVR_4	0xa503
+#define USB_DEVICE_ID_SAMSUNG_GEARVR_5	0xa504
+#define USB_DEVICE_ID_SAMSUNG_GEARVR_6	0xa505
+
 #endif
diff -Naur gearvr_kernel_removed/drivers/hid/hid-input.c gearvr_kernel/drivers/hid/hid-input.c
--- gearvr_kernel_removed/drivers/hid/hid-input.c	2016-04-10 00:15:02.959561900 -0500
+++ gearvr_kernel/drivers/hid/hid-input.c	2016-04-10 00:10:12.638922700 -0500
@@ -912,6 +912,17 @@
 		}
 		break;
 
+#ifdef CONFIG_USB_HMT_SAMSUNG_INPUT
+	case HID_UP_HMTVENDOR:
+		switch (usage->hid & HID_USAGE) {
+		case 0x001: map_key_clear(KEY_START_NOTA_CMD);		break;
+		case 0x002: map_key_clear(KEY_START_TA_CMD);		break;
+		case 0x003: map_key_clear(KEY_ONGOING_TA_CMD);		break;
+		default: goto ignore;
+		}
+		break;
+#endif
+
 	default:
 	unknown:
 		if (field->report_size == 1) {
diff -Naur gearvr_kernel_removed/drivers/hid/Makefile gearvr_kernel/drivers/hid/Makefile
--- gearvr_kernel_removed/drivers/hid/Makefile	2016-04-10 00:17:07.324141500 -0500
+++ gearvr_kernel/drivers/hid/Makefile	2016-04-10 00:10:12.583903400 -0500
@@ -98,6 +98,7 @@
 	hid-roccat-koneplus.o hid-roccat-konepure.o hid-roccat-kovaplus.o \
 	hid-roccat-lua.o hid-roccat-pyra.o hid-roccat-savu.o
 obj-$(CONFIG_HID_SAITEK)	+= hid-saitek.o
+obj-$(CONFIG_HID_OVR)		+= hid-ovr.o
 obj-$(CONFIG_HID_SAMSUNG)	+= hid-samsung.o
 obj-$(CONFIG_HID_SMARTJOYPLUS)	+= hid-sjoy.o
 obj-$(CONFIG_HID_SONY)		+= hid-sony.o
diff -Naur gearvr_kernel_removed/drivers/input/evdev.c gearvr_kernel/drivers/input/evdev.c
--- gearvr_kernel_removed/drivers/input/evdev.c	2016-04-10 00:17:40.793593100 -0500
+++ gearvr_kernel/drivers/input/evdev.c	2016-04-10 00:10:12.677415500 -0500
@@ -111,6 +111,11 @@
 		__pass_event(client, &event);
 		if (v->type == EV_SYN && v->code == SYN_REPORT)
 			wakeup = true;
+#ifdef CONFIG_USB_HMT_SAMSUNG_INPUT
+		if (v->type== EV_KEY && v->code >= KEY_HMT_CMD_START)
+			pr_info("%s type:KEY code:0x%x value:%x\n", __func__,
+					v->code, v->value);
+#endif
 	}
 
 	spin_unlock(&client->buffer_lock);
diff -Naur gearvr_kernel_removed/drivers/usb/notify/Kconfig gearvr_kernel/drivers/usb/notify/Kconfig
--- gearvr_kernel_removed/drivers/usb/notify/Kconfig	2016-04-10 00:20:34.577632300 -0500
+++ gearvr_kernel/drivers/usb/notify/Kconfig	2016-04-10 00:10:12.702914400 -0500
@@ -42,3 +42,11 @@
           This must to be defined in /usb/storage/Kconfig 
           directory originally. But this feature is defined
           in this Kconfig to gather samsung feature about usb host.
+
+config USB_HMT_SAMSUNG_INPUT
+	boolean "USB HMT inputs for samsung hmt devices"
+	depends on HID
+	help
+	  Some samsung smart phones support gear vr.
+          The samsung gear vr need some special hid input codes.
+          This feature enables special hid input codes.
diff -Naur gearvr_kernel_removed/include/linux/hid.h gearvr_kernel/include/linux/hid.h
--- gearvr_kernel_removed/include/linux/hid.h	2016-04-10 00:18:42.708632600 -0500
+++ gearvr_kernel/include/linux/hid.h	2016-04-10 00:10:12.730402600 -0500
@@ -168,6 +168,9 @@
 #define HID_UP_CUSTOM		0x00ff0000
 #define HID_UP_LOGIVENDOR	0xffbc0000
 #define HID_UP_SENSOR		0x00200000
+#ifdef CONFIG_USB_HMT_SAMSUNG_INPUT
+#define HID_UP_HMTVENDOR		0xff020000
+#endif
 
 #define HID_USAGE		0x0000ffff
 
@@ -702,6 +705,13 @@
 #define	PM_HINT_FULLON	1<<5
 #define PM_HINT_NORMAL	1<<1
 
+/* Applications from HID Usage Tables 4/8/99 Version 1.1 */
+/* We ignore a few input applications that are not widely used */
+#ifdef CONFIG_USB_HMT_SAMSUNG_INPUT
+#define IS_INPUT_APPLICATION(a) (((a >= 0x00010000) && (a <= 0x00010008)) || (a == 0x00010080) || (a == 0x000c0001) || ((a >= 0x000d0002) && (a <= 0x000d0006)) || a == 0xff020001)
+#else
+#define IS_INPUT_APPLICATION(a) (((a >= 0x00010000) && (a <= 0x00010008)) || (a == 0x00010080) || (a == 0x000c0001) || ((a >= 0x000d0002) && (a <= 0x000d0006)))
+#endif
 /* HID core API */
 
 extern int hid_debug;
diff -Naur gearvr_kernel_removed/include/uapi/linux/input.h gearvr_kernel/include/uapi/linux/input.h
--- gearvr_kernel_removed/include/uapi/linux/input.h	2016-04-10 00:19:20.189282000 -0500
+++ gearvr_kernel/include/uapi/linux/input.h	2016-04-10 00:10:12.759948000 -0500
@@ -804,6 +804,12 @@
 #define BTN_TRIGGER_HAPPY39		0x2e6
 #define BTN_TRIGGER_HAPPY40		0x2e7
 
+#ifdef CONFIG_USB_HMT_SAMSUNG_INPUT
+#define KEY_START_NOTA_CMD		0x2fc
+#define KEY_START_TA_CMD		0x2fd
+#define KEY_ONGOING_TA_CMD		0x2fe
+#define KEY_HMT_CMD_START		KEY_START_NOTA_CMD
+#endif
 
 /* We avoid low common keys in module aliases so they don't get huge. */
 #define KEY_MIN_INTERESTING	KEY_MUTE
diff -Naur gearvr_kernel_removed/README.md gearvr_kernel/README.md
--- gearvr_kernel_removed/README.md	1969-12-31 18:00:00.000000000 -0600
+++ gearvr_kernel/README.md	2016-04-10 00:10:12.533665700 -0500
@@ -0,0 +1,6 @@
+# gearvr_kernel
+
+GearVR kernel config options:
+
+CONFIG_HID_OVR=y
+CONFIG_USB_HMT_SAMSUNG_INPUT=y
